File: chapter_2_listproblem/Problem_03_RemoveNodeByRatio.cpp
#include <iostream>
#include <cmath>

using namespace std;

class Node{
public:
    Node(int data)
    {
        value = data;
    }

    ~Node(){}

    Node *next;
    int value;
};

Node* removeMidNode(Node* head)
{
    if(head == NULL || head->next == NULL)
    {
        return head;
    }
    if(head->next->next == NULL)
    {
        return head->next;
    }

    Node *pre = head;
    Node *cur = head->next->next;
    while(cur->next != NULL && cur->next->next != NULL)
    {
        pre = pre->next;
        cur = cur->next->next;
    }

    pre->next = pre->next->next;
    return head;
}

Node* removeByRatio(Node* head, int a, int b)
{
    if(a < 1 || a > b)
    {
        return head;
    }
    int n = 0;
    Node* cur = head;
    while(cur != NULL)
    {
        n++;
        cur = cur->next;
    }
    n = static_cast<int>(ceil(static_cast<double>(a * n) / static_cast<double>(b)));
    if(n == 1)
    {
        head = head->next;
    }
    if(n > 1)
    {
        cur = head;
        while(--n != 1)
        {
            cur = cur->next;
        }
        cur->next = cur->next->next;
    }
    return head;
}

void printLinkedList(Node *node)
{
    cout << "Linked List: ";
    while(node != NULL)
    {
        cout << node->value << " ";
        node = node->next;
    }
    cout << endl;
}

int main()
{
    Node *head = new Node(1);
    head->next = new Node(2);
    head->next->next = new Node(3);
    head->next->next->next = new Node(4);
    head->next->next->next->next = new Node(5);
    head->next->next->next->next->next = new Node(6);

    printLinkedList(head);
    head = removeMidNode(head);
    printLinkedList(head);
    head = removeByRatio(head, 2, 5);
    printLinkedList(head);
    head = removeByRatio(head, 1, 3);
    printLinkedList(head);

    return 0;
}

File: chapter_2_listproblem/Problem_03_RemoveNodeByRatio.java
package chapter_2_listproblem;

public class Problem_03_RemoveNodeByRatio {

	public static class Node {
		public int value;
		public Node next;

		public Node(int data) {
			this.value = data;
		}
	}

	public static Node removeMidNode(Node head) {
		if (head == null || head.next == null) {
			return head;
		}
		if (head.next.next == null) {
			return head.next;
		}
		Node pre = head;
		Node cur = head.next.next;
		while (cur.next != null && cur.next.next != null) {
			pre = pre.next;
			cur = cur.next.next;
		}
		pre.next = pre.next.next;
		return head;
	}

	public static Node removeByRatio(Node head, int a, int b) {
		if (a < 1 || a > b) {
			return head;
		}
		int n = 0;
		Node cur = head;
		while (cur != null) {
			n++;
			cur = cur.next;
		}
		n = (int) Math.ceil(((double) (a * n)) / (double) b);
		if (n == 1) {
			head = head.next;
		}
		if (n > 1) {
			cur = head;
			while (--n != 1) {
				cur = cur.next;
			}
			cur.next = cur.next.next;
		}
		return head;
	}

	public static void printLinkedList(Node head) {
		System.out.print("Linked List: ");
		while (head != null) {
			System.out.print(head.value + " ");
			head = head.next;
		}
		System.out.println();
	}

	public static void main(String[] args) {
		Node head = new Node(1);
		head.next = new Node(2);
		head.next.next = new Node(3);
		head.next.next.next = new Node(4);
		head.next.next.next.next = new Node(5);
		head.next.next.next.next.next = new Node(6);

		printLinkedList(head);
		head = removeMidNode(head);
		printLinkedList(head);
		head = removeByRatio(head, 2, 5);
		printLinkedList(head);
		head = removeByRatio(head, 1, 3);
		printLinkedList(head);

	}

}
